### PR TITLE
gender → sex

### LINKED ISSUE #207: gender → sex?
Although _sex_ might be perceived as offensive in other contexts, in the context of life sciences it must be used instead of _gender_, where appropriate. The [EDF+ specification](https://www.edfplus.info/specs/edfplus.html#additionalspecs) is clear about it:
> \- sex (English, so F or M).

Would you agree to fix this, if backwards compatibility is assured and tested, of course?

* [What are Sex & Gender? And why do they matter in health research?](https://orwh.od.nih.gov/sex-gender)
* [What Do We Mean By Sex and Gender?](https://medicine.yale.edu/news-article/what-do-we-mean-by-sex-and-gender/)

### PR DESCRIPTION
**Pull request by DimitriPapadopoulos**
Fixes #207 .

### CODE REVIEW

File: pyedflib/_extensions/_pyedflib.pyx [hunk 1]
```diff
@@ -270,8 +270,13 @@ cdef class CyEdfReader:
         def __get__(self):
             return self.hdr.patientcode
 
+    property sex:
+        def __get__(self):
+            return self.hdr.gender
+
     property gender:
         def __get__(self):
+            warnings.warn('This property is deprecated.', DeprecationWarning, stacklevel=2)
```
Comment (skjerns) [inline pyedflib/_extensions/_pyedflib.pyx:279]:
```suggestion
            warnings.warn('the `gender` property is deprecated. Please use `sex` instead.', DeprecationWarning, stacklevel=2)
```

File: pyedflib/_extensions/_pyedflib.pyx [hunk 2]
```diff
@@ -615,10 +620,14 @@ def rewind(handle, edfsignal):
...
+def set_sex(handle, sex):
+    """int edf_set_sex(int handle, int sex)"""
+    if sex is None: return 0 #don't set sex at all to prevent default 'F'
+    return c_edf.edf_set_gender(handle, sex)
+
 def set_gender(handle, gender):
-    """int edf_set_gender(int handle, int gender)"""
-    if gender is None: return 0 #don't set gender at all to prevent default 'F'
-    return c_edf.edf_set_gender(handle, gender)
+    warnings.warn('This function is deprecated.', DeprecationWarning, stacklevel=2)
```
Comment (skjerns) [inline pyedflib/_extensions/_pyedflib.pyx:629]:
```suggestion
    warnings.warn('`set_gender` is deprecated, use `set_sex` instead.', DeprecationWarning, stacklevel=2)
```

File: pyedflib/_extensions/_pyedflib.pyx [hunk 3]
```diff
@@ -615,10 +620,14 @@ def rewind(handle, edfsignal):
...
+    set_sex(handle, gender)
```
Comment (skjerns) [inline pyedflib/_extensions/_pyedflib.pyx:630]:
```suggestion
    return set_sex(handle, gender)
```

File: pyedflib/edfwriter.py [hunk 4]
```diff
@@ -126,17 +126,22 @@ def isbytestr(s):
...
-    elif gender.lower() in  ["male", "man", "m"]:
+    elif sex.lower() in  ["male", "man", "m"]:
         return 1
     else:
-        raise ValueError(f"Unknown gender: '{gender}'")
+        raise ValueError(f"Unknown sex: '{sex}'")
+
+
+def gender2int(sex):
+    warnings.warn('This function is deprecated.', DeprecationWarning, stacklevel=2)
```
Comment (skjerns) [inline pyedflib/edfwriter.py:143]:
```suggestion
    warnings.warn('`gender2int`  is deprecated, use `sex2int` instead.', DeprecationWarning, stacklevel=2)
```

File: pyedflib/edfwriter.py [hunk 5]
```diff
@@ -126,17 +126,22 @@ def isbytestr(s):
...
+    sex2int(sex)
```
Comment (skjerns) [inline pyedflib/edfwriter.py:144]:
```suggestion
    return sex2int(sex)
```

File: pyedflib/edfwriter.py [hunk 6]
```diff
@@ -440,17 +445,17 @@ def setAdmincode(self, admincode):
         self.admincode = admincode
         self.update_header()
 
-    def setGender(self, gender):
+    def setSex(self, sex):
         """
-        Sets the gender.
+        Sets the sex.
```
Comment (skjerns) [inline pyedflib/edfwriter.py:450]:
```suggestion
        Sets the sex. Due to the edf specifications, only binary assignment is possible.
```

File: pyedflib/highlevel.py [hunk 7]
```diff
@@ -191,13 +192,23 @@ def make_header(technician='', recording_additional='', patientname='',
...
         birthdate = birthdate.strftime('%d %b %Y').lower()
+
+    # backwards compatibility
+    if gender is not None:
+        if sex == '':
+            sex = gender
+        elif gender != sex:
+            raise ValueError(f'Passed both sex and gender, with different values: "{sex}" != "{gender}"')
+    del gender
+
```
Comment (skjerns) [inline pyedflib/highlevel.py:203]: give deprecation warning is gender is set instead of sex

**Comment by DimitriPapadopoulos** [comment]
Ouch. Somehow I lost the _gender_ branch. Will retry.

File: pyedflib/edfwriter.py (hunk 6 above)
Comment (DimitriPapadopoulos) [reply pyedflib/edfwriter.py:449]: I could add that, but:
- that was already true before this PR,
-  I don't see the need to justify ourselves, the EDF library implements the EDF specs as a matter of course.

File: pyedflib/highlevel.py (hunk 7 above)
Comment (DimitriPapadopoulos) [reply pyedflib/highlevel.py]: Done.

File: pyedflib/edfwriter.py (hunk 5 above)
Comment (DimitriPapadopoulos) [reply pyedflib/edfwriter.py]: That error is gone in the new version written from scratch :grin:

File: pyedflib/_extensions/_pyedflib.pyx (hunk 3 above)
Comment (DimitriPapadopoulos) [reply pyedflib/_extensions/_pyedflib.pyx]: hat error is gone in the new version written from scratch grin

File: pyedflib/edfwriter.py (hunk 4 above)
Comment (DimitriPapadopoulos) [reply pyedflib/edfwriter.py]: Done.

File: pyedflib/_extensions/_pyedflib.pyx (hunk 2 above)
Comment (DimitriPapadopoulos) [reply pyedflib/_extensions/_pyedflib.pyx]: Done.

File: pyedflib/_extensions/_pyedflib.pyx (hunk 1 above)
Comment (DimitriPapadopoulos) [reply pyedflib/_extensions/_pyedflib.pyx]: Done.

**Comment by skjerns** [comment]
Let me know when you are done then I'll review.

Also: Could you sqash the commits after finishing?

**Comment by DimitriPapadopoulos** [comment]
The new code should be backwards compatible enough. Not sure if I test the backwards compatibility stuff enough.

File: pyedflib/edfwriter.py (hunk 6 above)
Comment (skjerns) [reply pyedflib/edfwriter.py:449]: true, but people do not read the specs, so I'd like to have it added

**Comment by DimitriPapadopoulos** [comment]
About "_Due to the edf specifications, only binary assignment is possible_": I am not a specialist, but I understand specialists describe biological sex as binary, while fully endorsing gender diversity. See for example  https://doi.org/10.1002/bies.202200173.

**Comment by DimitriPapadopoulos** [comment]
We do have such a test in [`pyedflib/tests/test_edfreader.py`](https://github.com/DimitriPapadopoulos/pyedflib/blob/0bf716563385ae9202f6d5dc8ddb63c768f3349a/pyedflib/tests/test_edfreader.py#L181):
```python
        np.testing.assert_equal(f.getSex(), 'Male')
        np.testing.assert_equal(f.getGender(), 'Male')  # deprecated
```
I have just added tests to `pyedflib/tests/test_edfwriter.py` in `test_sex_setting_correctly()`.

**Comment by skjerns** [comment]
perfect, I think it's ready to merge now :)

**Pull request merged**